packages/react-client/src/ReactFlightClient.js

```diff
@@ -141,28 +145,81 @@ function definePendingProperty(
 }
 
+function createElement(type, key, props): React$Element<any> {
+  const element: any = {
+    // This tag allows us to uniquely identify this as a React Element
+    $$typeof: REACT_ELEMENT_TYPE,
+
+    // Built-in properties that belong on the element
+    type: type,
+    key: key,
+    ref: null,
+    props: props,
+
+    // Record the component responsible for creating this element.
+    _owner: null,
+  };
+  if (__DEV__) {
+    // We don't really need to add any of these but keeping them for good measure.
+    // Unfortunately, _store is enumerable in jest matchers so for equality to
+    // work, I need to keep it or make _store non-enumerable in the other file.
+    element._store = {};
+    Object.defineProperty(element._store, 'validated', {
+      configurable: false,
+      enumerable: false,
+      writable: true,
+      value: true, // This element has already been validated on the server.
+    });
+    Object.defineProperty(element, '_self', {
+      configurable: false,
+      enumerable: false,
+      writable: false,
+      value: null,
+    });
+    Object.defineProperty(element, '_source', {
+      configurable: false,
+      enumerable: false,
+      writable: false,
+      value: null,
+    });
+  }
+  return element;
+}
+
 export function parseModelFromJSON(
   response: Response,
   targetObj: Object,
   key: string,
   value: JSONValue,
-): any {
-  if (typeof value === 'string' && value[0] === '$') {
-    if (value[1] === '$') {
-      // This was an escaped string value.
-      return value.substring(1);
-    } else {
-      let id = parseInt(value.substring(1), 16);
-      let chunks = response.chunks;
-      let chunk = chunks.get(id);
-      if (!chunk) {
-        chunk = createPendingChunk();
-        chunks.set(id, chunk);
-      } else if (chunk.status === RESOLVED) {
-        return chunk.value;
+): mixed {
+  if (typeof value === 'string') {
+    if (value[0] === '$') {
+      if (value === '$') {
+        return REACT_ELEMENT_TYPE;
+      } else if (value[1] === '$' || value[1] === '@') {
+        // This was an escaped string value.
+        return value.substring(1);
+      } else {
+        let id = parseInt(value.substring(1), 16);
+        let chunks = response.chunks;
+        let chunk = chunks.get(id);
+        if (!chunk) {
+          chunk = createPendingChunk();
+          chunks.set(id, chunk);
+        } else if (chunk.status === RESOLVED) {
+          return chunk.value;
+        }
+        definePendingProperty(targetObj, key, chunk);
+        return undefined;
       }
-      definePendingProperty(targetObj, key, chunk);
-      return undefined;
+    }
+  }
+  if (isArray(value)) {
+    let tuple: [mixed, mixed, mixed, mixed] = (value: any);
+    if (tuple[0] === REACT_ELEMENT_TYPE) {
+      // TODO: Consider having React just directly accept these arrays as elements.
+      // Or even change the ReactElement type to be an array.
+      return createElement(tuple[1], tuple[2], tuple[3]);
```

Comment:
With that change we'd just be able to pass arrays through without deserializing, right? Though we'd still call this function on the props element, right? Or I guess those are flattened(?). The overall architecture makes sense but i'm still catching up on the details of each stage in the expansion, serialization, deserialization, and rendering.

Reply:
There is still a deserialization step for the strings. String values starting with `$` have special meaning so we need to deserialize those but we also need to unescape the user space strings that start with `$`.

JSON.parse materializes all objects and arrays in the JSON and then traverses over them by mutating them with new values. Which we'd still need to do to replace `"$"` with the symbol and replace references to reused or future values (i.e. "$12345" to their materialized value).

The benefit would be that we'd reuse the array already created by JSON.parse and just mutate the first slot from a string to a symbol.

We need the symbol to avoid the XSS potential in https://github.com/facebook/react/issues/3473.

The problem is that if this value gets passed into a render function and then returned from the render, we don't know if that came from Flight on the server or user provided JSON.

```
props: { userdata: ["$", "script", {children: "alert('p0wned')"}] }
```

So I solve this by escaping the `"$"` to `"$$"` in user provided data. At some point, I need to unescape it though. That's where deserialization comes in.

It's also used for references and I expect us to use it for more things.

In the raw Flight streaming model, this is done using the second argument to JSON.parse so it's kind of inline with the parser which (I hope) is very fast.

The issue with the Relay layering is that we currently have to do a second traversal. We could possibly avoid that if we were hooked in somehow to the same JSON.parse that Relay uses or any other pre-existing traversals.